lib/src/main/java/at/bitfire/cert4android/UserDecisionRegistry.kt

```diff
@@ -33,139 +32,76 @@ class UserDecisionRegistry private constructor(
+     *   - cancelled when the scope is cancelled.
+     */
+    internal val pendingDecisions = Collections.synchronizedMap(mutableMapOf<X509Certificate, MutableList<Continuation<Boolean>>>())
 
     /**
      * Tries to retrieve a trust decision from the user about a given certificate.
      *
      * Thread-safe, can handle multiple requests for various certificates and/or the same certificate at once.
      *
      * @param cert              certificate to ask user about
-     * @param appInForeground   whether the app is currently in foreground = whether it can directly launch an Activity
-     * @return *true* if the user explicitly trusts the certificate, *false* if unknown or untrusted
+     * @param scope             the coroutine scope within which [getUserDecision] is executed
+     * @param getUserDecision   suspendable function to retrieve user decision
+     * @return                  *true* if the user explicitly trusts the certificate; *false* if
+     *                          unknown or untrusted
      */
-    suspend fun check(cert: X509Certificate, appInForeground: Boolean): Boolean = suspendCancellableCoroutine { cont ->
-        // check whether we're able to retrieve user feedback (= start an Activity and/or show a notification)
-        val notificationsPermitted = NotificationUtils.notificationsPermitted(context)
-        val userDecisionPossible = appInForeground || notificationsPermitted
-
-        if (userDecisionPossible) {
-            // User decision possible → remember request in pendingDecisions so that a later decision will be applied to this request
-
-            cont.invokeOnCancellation {
-                // remove from pending decisions on cancellation
-                synchronized(pendingDecisions) {
-                    pendingDecisions[cert]?.remove(cont)
-                }
-
-                val nm = NotificationUtils.createChannels(context)
-                nm.cancel(CertUtils.getTag(cert), NotificationUtils.ID_CERT_DECISION)
-            }
-
-            val requestDecision: Boolean
-            synchronized(pendingDecisions) {
-                if (pendingDecisions.containsKey(cert)) {
-                    // There are already pending decisions for this request, just add our request
-                    pendingDecisions[cert]!! += cont
-                    requestDecision = false
-                } else {
-                    // First decision for this certificate, show UI
-                    pendingDecisions[cert] = mutableListOf(cont)
-                    requestDecision = true
-                }
-            }
+    suspend fun check(
+        cert: X509Certificate,
+        scope: CoroutineScope,
+        getUserDecision: suspend (X509Certificate) -> Boolean
+    ): Boolean = suspendCancellableCoroutine { cont ->
+        cont.invokeOnCancellation {
+            // remove from pending decisions on cancellation
+            pendingDecisions[cert]?.remove(cont)
```

Comment:
What happens in the case the the cancelled Continuation was the last one in the list?

Then `pendingDecisions[cert]` will forever remain an empty list, right? That would also mean that the UI (`getUserDecision()`) will not be called again.

This may even be related to https://github.com/bitfireAT/davx5-ose/discussions/861

I wonder whether we could find a clearer way to handle pending decisions.

Reply:
Hm, maybe it was intended like that…

Reply:
I think I misunderstood something. To be clear:

> What happens in the case the the cancelled Continuation was the last one in the list?
> 
> Then `pendingDecisions[cert]` will forever remain an empty list, right? 

Correct. The map entry will remain until `pendingDecisions` is recreated within `UserDecisionRegistry`.

> That would also mean that the UI (`getUserDecision()`) will not be called again. 

Not sure. I don't see how yet.

> This may even be related to [bitfireAT/davx5-ose#861](https://github.com/bitfireAT/davx5-ose/discussions/861)

maybe ...

Reply:
Having looked at it more in-depth now. I am quite sure you are right. The UI (`getUserDecision()`) will not be called again if `pendindDecision[cert]` still contains an empty list (the map entry not removed or set to null). 

I have added a test and the conditional removal.